lib/Cmts.ml

```diff
@@ -16,14 +16,109 @@ open Migrate_ast
 open Asttypes
 open Parsetree
 
+module Location_comp = struct
+  include Location
+  include Comparator.Make (Location)
+end
+
+module Loc_map : sig
+  type 'a t
+
+  val empty : 'a t
+
+  val add_list : 'a t -> Location.t -> 'a list -> 'a t
+
+  val update_multi :
+       'a t
+    -> src:Location.t
+    -> dst:Location.t
+    -> f:('a list -> 'a list -> 'a list)
+    -> 'a t
+
+  val find : 'a t -> Location.t -> 'a list option
+
+  val remove : 'a t -> Location.t -> 'a t
+
+  val filter : 'a t -> f:('a -> bool) -> 'a t
+
+  val mem : 'a t -> Location.t -> bool
+
+  val to_list : 'a t -> 'a list
+
+  val find_multi : 'a t -> Location.t -> 'a list
+end = struct
+  type 'a t = (Location.t, 'a list, Location_comp.comparator_witness) Map.t
+
+  let add_list map key data = Map.add_exn map ~key ~data
+
+  let remove map loc = Map.remove map loc
+
+  let update_multi map ~src ~dst ~f =
+    Option.fold (Map.find map src) ~init:(Map.remove map src)
+      ~f:(fun new_map src_data ->
+        Map.update new_map dst ~f:(fun dst_data ->
+            Option.fold dst_data ~init:src_data ~f))
+
+  let empty = Map.empty (module Location_comp)
+
+  let find map loc = Map.find map loc
+
+  let filter map ~f = Map.map map ~f:(List.filter ~f)
+
+  let mem map loc = Map.mem map loc
+
+  let to_list map = Map.to_alist map |> List.concat_map ~f:snd
+
+  let find_multi map loc = Map.find_multi map loc
+end
+
+module Loc_set : sig
+  type t
+
+  val empty : t
+
+  val add : Location.t -> t -> t
+
+  val remove : Location.t -> t -> t
+
+  val to_list : t -> Location.t list
+end = struct
+  type t = (Location.t, Location_comp.comparator_witness) Set.t
+
+  let empty = Set.empty (module Location_comp)
+
+  let add key t = Set.add t key
+
+  let remove key t = Set.remove t key
+
+  let to_list t = Set.to_list t
+end
+
```

Comment:
Would it make sense to move these modules to `Cmt`, e.g. `Cmt.Set` and `Cmt.Map` to remove some weight from this file?

Reply:
yes, I think that would work. Map is a multimap (it returns lists) so the name can reflect that as well. `Location_comp` can also be merged into `Location`.

Reply:
They can even end up in their own files if we want to keep them independent from comments, or even in a utility file.

Reply:
I just did this. Actually, these are *location* set/multimaps, not *cmt* ones, so I moved them into `Location`. It wasn't to clear what `Migrate_ast` was exposing, so I also added an mli file for this module (spoiler: there was minor bits of dead code).

Reply:
The mli file I wrote was specific to 4.08, I'll fix it